### PR TITLE
fix: error IAM role attachement when applying the module the first ti…

### LINKED ISSUE #591: Error putting IAM role policy, and Error attaching policy on creation
I'm using version `5.5.0` of the `npalm/gitlab-runner/aws` module.

When instantiating a `npalm/gitlab-runner/aws` module, almost all resources are getting created, except for these two:

```tf
Error: Error putting IAM role policy gitlab-runners-instance-role: NoSuchEntity: The role with name gitlab-runners-instance cannot be found. status code: 404, request id: b86fa6e7-1f54-47c2-89d2-020f7c747c81
with module.gitlab_runner.aws_iam_role_policy.instance[0]
on .terraform/modules/gitlab_runner/logging.tf line 1, in resource "aws_iam_role_policy" "instance":
resource "aws_iam_role_policy" "instance" {
```

```tf
Error: Error attaching policy arn:aws:iam::091232145884:policy/amberatung-gitlab-runners-docker-machine-cache to IAM Role gitlab-runners-instance: NoSuchEntity: The role with name gitlab-runners-instance cannot be found. status code: 404, request id: 65e813ca-5b00-4033-b2bc-22bcf20f375a
with module.gitlab_runner.aws_iam_role_policy_attachment.docker_machine_cache_instance[0]
on .terraform/modules/gitlab_runner/main.tf line 399, in resource "aws_iam_role_policy_attachment" "docker_machine_cache_instance":
resource "aws_iam_role_policy_attachment" "docker_machine_cache_instance" {
```

Yet, none of these errors come up on the second run, and the plan gets applied successfully.

### PR DESCRIPTION
**Pull request by npalm**
## Description

fix #591

Ensure policies attachment are done with module managed resources in case the role is managed via terraform objects. This to avoid attachments are created before the role is created. Depends on does not support dynamics, sot not an option.

## Migrations required

NO

## Verification

- [x] default example

### CODE REVIEW

**Comment by npalm** [comment]
@kayman-mk Tried to fix the issue, but the docker machine is not booting yet. I have not more time to dig in, will have a look later. Or feel free to dig in as well

**Comment by npalm** [comment]
@kayman-mk would you the option to rebase this PR

**Comment by kayman-mk** [comment]
Merge conflicts solved.

I changed some more policies and tested the change. First apply succeeded. 👍

File: logging.tf [hunk 1]
```diff
@@ -1,7 +1,7 @@
 resource "aws_iam_role_policy" "instance" {
   count  = var.enable_cloudwatch_logging && var.create_runner_iam_role ? 1 : 0
   name   = "${local.name_iam_objects}-logging"
-  role   = local.aws_iam_role_instance_name
+  role   = var.create_runner_iam_role ? aws_iam_role.instance[0].name : var.runner_iam_role_name
```
Comment (kayman-mk) [inline logging.tf:4]: suggestion: replace `var.runner_iam_role_name`by `local.aws_iam_role_instance_name` in all changes.

The local variable respects the `var.overrides["name_iam_objects"]` which was introduced to align the names of IAM objects. The variable description says: `If unspecified then `{name_iam_objects}-instance` is used`.

**Review by kayman-mk — CHANGES_REQUESTED** [review]

File: main.tf [hunk 2]
```diff
@@ -396,14 +396,14 @@ resource "aws_iam_policy" "instance_session_manager_policy" {
 resource "aws_iam_role_policy_attachment" "instance_session_manager_policy" {
   count = var.enable_runner_ssm_access ? 1 : 0
 
-  role       = local.aws_iam_role_instance_name
+  role       = var.create_runner_iam_role ? aws_iam_role.instance[0].name : var.runner_iam_role_name
```
Comment (kayman-mk) [inline main.tf:399]:
```suggestion
  role       = var.create_runner_iam_role ? aws_iam_role.instance[0].name : local.aws_iam_role_instance_name
```

File: logging.tf (hunk 1 above)
Comment (kayman-mk) [inline logging.tf:4]:
```suggestion
  role   = var.create_runner_iam_role ? aws_iam_role.instance[0].name : local.aws_iam_role_instance_name
```

File: main.tf [hunk 3]
```diff
@@ -396,14 +396,14 @@ resource "aws_iam_policy" "instance_session_manager_policy" {
...
   policy_arn = aws_iam_policy.instance_session_manager_policy[0].arn
 }
 
 resource "aws_iam_role_policy_attachment" "instance_session_manager_aws_managed" {
   count = var.enable_runner_ssm_access ? 1 : 0
 
-  role       = local.aws_iam_role_instance_name
+  role       = var.create_runner_iam_role ? aws_iam_role.instance[0].name : var.runner_iam_role_name
```
Comment (kayman-mk) [inline main.tf:406]:
```suggestion
  role       = var.create_runner_iam_role ? aws_iam_role.instance[0].name : local.aws_iam_role_instance_name
```

File: main.tf [hunk 4]
```diff
@@ -413,7 +413,7 @@ resource "aws_iam_role_policy_attachment" "instance_session_manager_aws_managed"
 resource "aws_iam_role_policy_attachment" "user_defined_policies" {
   count = length(var.runner_iam_policy_arns)
 
-  role       = local.aws_iam_role_instance_name
+  role       = var.create_runner_iam_role ? aws_iam_role.instance[0].name : var.runner_iam_role_name
```
Comment (kayman-mk) [inline main.tf:416]:
```suggestion
  role       = var.create_runner_iam_role ? aws_iam_role.instance[0].name : local.aws_iam_role_instance_name
```

File: main.tf [hunk 5]
```diff
@@ -426,7 +426,7 @@ resource "aws_iam_role_policy_attachment" "docker_machine_cache_instance" {
      use aws_iam_role.docker_machine.name here! See https://docs.gitlab.com/runner/configuration/advanced-configuration.html */
   count = var.runners_executor == "docker+machine" ? (var.cache_bucket["create"] || lookup(var.cache_bucket, "policy", "") != "" ? 1 : 0) : 0
 
-  role       = local.aws_iam_role_instance_name
+  role       = var.create_runner_iam_role ? aws_iam_role.instance[0].name : var.runner_iam_role_name
```
Comment (kayman-mk) [inline main.tf:429]:
```suggestion
  role       = var.create_runner_iam_role ? aws_iam_role.instance[0].name : local.aws_iam_role_instance_name
```

File: main.tf [hunk 6]
```diff
@@ -482,7 +482,7 @@ resource "aws_iam_policy" "service_linked_role" {
 resource "aws_iam_role_policy_attachment" "service_linked_role" {
   count = var.allow_iam_service_linked_role_creation ? 1 : 0
 
-  role       = local.aws_iam_role_instance_name
+  role       = var.create_runner_iam_role ? aws_iam_role.instance[0].name : var.runner_iam_role_name
```
Comment (kayman-mk) [inline main.tf:485]:
```suggestion
  role       = var.create_runner_iam_role ? aws_iam_role.instance[0].name : local.aws_iam_role_instance_name
```

File: main.tf [hunk 7]
```diff
@@ -506,7 +506,7 @@ resource "aws_iam_policy" "ssm" {
 resource "aws_iam_role_policy_attachment" "ssm" {
   count = var.enable_manage_gitlab_token ? 1 : 0
 
-  role       = local.aws_iam_role_instance_name
+  role       = var.create_runner_iam_role ? aws_iam_role.instance[0].name : var.runner_iam_role_name
```
Comment (kayman-mk) [inline main.tf:509]:
```suggestion
  role       = var.create_runner_iam_role ? aws_iam_role.instance[0].name : local.aws_iam_role_instance_name
```

File: main.tf [hunk 8]
```diff
@@ -526,7 +526,7 @@ resource "aws_iam_policy" "eip" {
 resource "aws_iam_role_policy_attachment" "eip" {
   count = var.enable_eip ? 1 : 0
 
-  role       = local.aws_iam_role_instance_name
+  role       = var.create_runner_iam_role ? aws_iam_role.instance[0].name : var.runner_iam_role_name
```
Comment (kayman-mk) [inline main.tf:529]:
```suggestion
  role       = var.create_runner_iam_role ? aws_iam_role.instance[0].name : local.aws_iam_role_instance_name
```

**Review by kayman-mk — CHANGES_REQUESTED** [review]

**Comment by npalm** [comment]
@kayman-mk accepted your suggestion, PR is good to go. Please can you review the PR?

**Comment by npalm** [comment]
@kayman-mk can you review the PR, should be ready to merge

**Review by kayman-mk — APPROVED** [review]

**Pull request merged**